icu4j/main/classes/core/src/com/ibm/icu/text/ListFormatter.java

```diff
@@ -526,14 +530,127 @@ FormattedListBuilder formatImpl(Collection<?> items, boolean needsFields) {
+    // handle the the two and end pattern.
+    private class ContextualHandler implements PatternHandler {
+        ContextualHandler(Pattern regexp, String thenTwo, String elseTwo, String thenEnd, String elseEnd) {
+            this.regexp = regexp;
+            this.thenTwoPattern = thenTwo;
+            this.elseTwoPattern = elseTwo;
+            this.thenEndPattern = thenEnd;
+            this.elseEndPattern = elseEnd;
+        }
+
+        public String getTwoPattern(String text) {
+            if(this.regexp.matcher(text).matches()) {
+                return this.thenTwoPattern;
+            } else {
+                return this.elseTwoPattern;
+            }
+        }
+
+        public String getEndPattern(String text) {
+            if(this.regexp.matcher(text).matches()) {
+                return this.thenEndPattern;
+            } else {
+                return this.elseEndPattern;
+            }
+        }
+
+        private final Pattern regexp;
+        private final String thenTwoPattern;
+        private final String elseTwoPattern;
+        private final String thenEndPattern;
+        private final String elseEndPattern;
+
+    }
+
+    // Pattern in the ICU Data which might be replaced y by e.
+    private static final String compiledY = compilePattern("{0} y {1}", new StringBuilder());
+
+    // The new pattern to replace y to e
+    private static final String compiledE = compilePattern("{0} e {1}", new StringBuilder());
+
+    // Pattern in the ICU Data which might be replaced o by u.
+    private static final String compiledO = compilePattern("{0} o {1}", new StringBuilder());
+
+    // The new pattern to replace u to o
+    private static final String compiledU = compilePattern("{0} u {1}", new StringBuilder());
+
+    // Condiction to change to e.
+    // Starts with "hi" or "i" but not with "hie" nor "hia"a
+    private static final Pattern changeToE = Pattern.compile("(?i)^(i|hi[^ae]).*$");
+
+    // Condiction to change to u.
+    // Starts with "o", "ho", and "8". Also "11" by itself.
+    private static final Pattern changeToU = Pattern.compile("(?i)^((o|ho|8).*|11)$");
+
+    // Pattern in the ICU Data which might need to add a DASH after VAV.
+    private static final String compiledVav = compilePattern("{0} \u05D5{1}", new StringBuilder());
+
+    // Pattern to add a DASH after VAV.
+    private static final String compiledVavDash = compilePattern("{0} \u05D5-{1}", new StringBuilder());
+
+    // Condiction to change to VAV follow by a dash.
+    // Starts with non Hebrew letter.
+    private static final Pattern changeToVavDash = Pattern.compile("^[\\P{InHebrew}].*$");
+
+    // A factory function to create function based on locale
+    // Handle specal case of Spanish and Hebrew
+    private PatternHandler createPatternHandler(String two, String end) {
+        if (this.locale != null) {
+            String language = this.locale.getLanguage();
+            if (language == "es") {
```

Comment:
You can't `==` strings in Java.  But, in that case, how are the tests passing?

Reply:
good catch- "es" is string literal and getLanguage probably return the same from the string pool. I will change it.